kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -705,6 +724,52 @@ protected boolean checkAuthzGroups(String userId, String function, String entity
 			azgs = ref.getAuthzGroups(userId);
 		}
 
+		// need to know whether role swap is in effect, since we can't share the cache entry between sessions
+		// that are swapped and not swapped
+
+		String siteRef = null;
+		String roleswap = null;
+
+		// Actual code in DbAuthzGroupService will not roleswap if there's a user site ref in the list and
+		// it is acceptable. However we can't tell that without doing a database access, so be conservative
+		// and cache it separately as a role swap
+		// This code does not handle delegated access
+		if (azgs != null && userId != null && userId.equals(sessionManager().getCurrentSessionUserId())) {
```

Comment:
The role swap should be site based... So I think the current way does make sense though I agree with you in that it should be something simpler like
ROLESWAP_PREFIX + /site/mercury
and not contain any other reference types. I had also done this same thing in my PR.